### ISSUE TITLE
cpu

### ISSUE THREAD

**Issue by mishav78** [mowshon/age-and-gender#5]
will it work with cpu only?

**Comment by mowshon** [comment]
@mishav78 Yes, it runs on CPU only, and that's intentional. You don't need a GPU or CUDA.

- Age and gender run on [ONNX Runtime](https://onnxruntime.ai/) with the `CPUExecutionProvider`. The package depends on plain `onnxruntime`, not `onnxruntime-gpu`.
- Face detection (dlib's HOG detector) and the 5-point landmark model come from the prebuilt `dlib-bin` wheel, which is CPU-only.
- The models are small (a ResNet-10 on 64×64 crops for age, a compact CNN on 32×32 crops for gender), so they're fast on a CPU: about 2 ms per face for age and under 1 ms for gender.

The CPU provider is fixed on purpose. The ONNX models were checked against the original dlib networks on CPU, so the package refuses to build a session on any other provider.

### Running the models on a GPU anyway (unsupported)

If you have an NVIDIA GPU, you can swap the ONNX Runtime sessions yourself. Setting `providers` or installing `onnxruntime-gpu` isn't enough, because the package always asks for the CPU provider.

1. Replace `onnxruntime` with the GPU build. The `[cuda,cudnn]` extra installs the CUDA and cuDNN libraries from pip:

   ```bash
   pip uninstall -y onnxruntime
   pip install "onnxruntime-gpu[cuda,cudnn]"
   ```

2. Replace the sessions after you create the predictor:

   ```python
   import onnxruntime as ort
   from PIL import Image
   from age_and_gender import AgeAndGender

   ort.preload_dlls()  # load CUDA/cuDNN from the pip-installed nvidia-* wheels

   predictor = AgeAndGender()
   for network in predictor._engine._networks.values():
       network._session = ort.InferenceSession(
           predictor._bundle.model_bytes(network.task),
           providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
       )

   image = Image.open("photo.jpg").convert("RGB")
   print(predictor.predict(image))
   ```

   Check that the GPU is being used with `network._session.get_providers()`. The list should start with `CUDAExecutionProvider`.

Things to know before you do this:

- **This uses private attributes** (`_engine`, `_networks`, `_session`, `_bundle`). They can change in any release without notice.
- **The results are close to the CPU results but not identical.** On an RTX 3090, the probabilities differed from the CPU by up to about 0.005. Every predicted class was the same, but an `age` value or a `confidence` percentage can sometimes differ by 1.
- **`predict()` barely gets faster.** Only the age and gender networks move to the GPU. Face detection and landmarks still run on the CPU in dlib, and that's where most of the time goes. In our test, the networks got about 35× faster (age: 2.2 → 0.06 ms per face, gender: 0.7 → 0.02 ms per face, in batches of 256). But `predict()` on a photo with 5 faces only went from about 86 ms to about 75 ms.

The GPU is worth it mainly if you use your own (GPU) face detector and pass many cropped faces to `predict_face()`, `age()` or `gender()`. For normal use, the CPU build is fast enough.
